rust/feature-flags/src/flags/flag_matching.rs

```diff
@@ -454,6 +474,80 @@ impl FeatureFlagMatcher {
         }
     }
 
+    /// Evaluates and caches static cohort memberships for the current person.
+    /// This should be called once per request to avoid multiple DB lookups.
+    async fn evaluate_and_cache_static_cohorts(
+        &mut self,
+        cohorts: &[Cohort],
+    ) -> Result<HashMap<CohortId, bool>, FlagError> {
+        // Skip if we've already cached the results
+        if self.flag_evaluation_state.static_cohort_matches.is_some() {
+            return Ok(self
+                .flag_evaluation_state
+                .static_cohort_matches
+                .clone()
+                .unwrap());
+        }
+
+        let person_id = self.get_person_id().await?;
+        let static_cohorts: Vec<_> = cohorts.iter().filter(|c| c.is_static).collect();
+
+        if static_cohorts.is_empty() {
+            // Cache empty map to indicate we've checked
+            self.flag_evaluation_state.static_cohort_matches = Some(HashMap::new());
+            return Ok(HashMap::new());
+        }
+
+        let results = evaluate_static_cohorts(
+            self.reader.clone(),
+            person_id,
+            static_cohorts.iter().map(|c| c.id).collect(),
+        )
+        .await?
+        .into_iter()
+        .collect::<HashMap<_, _>>();
+
+        self.flag_evaluation_state.static_cohort_matches = Some(results.clone());
+        Ok(results.clone())
+    }
+
+    /// Evaluates cohort filters using cached static cohort results where possible.
+    /// For dynamic cohorts, evaluates them based on the provided properties.
+    pub async fn evaluate_cohort_filters(
+        &mut self,
+        cohort_property_filters: &[PropertyFilter],
+        target_properties: &HashMap<String, Value>,
+    ) -> Result<bool, FlagError> {
+        // At the start of the request, fetch all of the cohorts for the project from the cache
+        let cohorts = self.cohort_cache.get_cohorts(self.project_id).await?;
+
+        // Get cached static cohort results or evaluate them if not cached
+        let static_cohort_matches = match self.flag_evaluation_state.static_cohort_matches.as_ref()
+        {
+            Some(matches) => matches.clone(),
+            None => self.evaluate_and_cache_static_cohorts(&cohorts).await?,
```

Comment:
this shouldn't ever happen, but if it does it will happen once and then ideally never again.  Much better than before.